### PR TITLE
feat(py-client): Support creating pre-signed URLs

### PR DESCRIPTION
**Pull request by lcian**
Adds `Session.presigned_object_url("GET" | "HEAD", key, duration)` to the Python client, so a service holding an EdDSA keypair can hand out a time-limited URL that authorizes a single GET/HEAD without an auth token.

The subtle part is encoding. The server verifies the signature against the raw, still percent-encoded path and query it receives, so the client must sign the exact bytes its HTTP client puts on the wire.
urllib3 percent-encodes request targets via `urllib3.util.url._encode_target` in `HTTPConnectionPool.urlopen`, so rather than reimplement RFC 3986 and risk drift, `presign.py` reuses those same urllib3 internals — the match then holds by construction.
A unit test bench asserts parity with urllib3's wire encoding across corner cases, and end-to-end tests fetch the signed URLs against a live server with stdlib `urllib` to prove they work for external consumers.

Breaking change: `TokenGenerator` was renamed to `SecretKey` because with pre-signed URL support, the key is no longer just a JWT generator — it's a general-purpose EdDSA signing identity used for both JWT auth tokens and URL signatures. Related rename: `sign_for_scope()` → `token_for_scope()` (clarifies it returns a JWT token).
I've considered keeping `TokenGenerator` as a deprecated alias for `SecretKey`, but IMO we might as well remove it directly, as migration is trivial, and we'll be the ones doing the migration in Sentry itself.

Refs FS-386

### CODE REVIEW

File: clients/python/src/objectstore_client/client.py [hunk 1]
```diff
@@ -446,6 +447,78 @@ def object_url(self, key: str) -> str:
...
+        key: str,
+        duration: timedelta,
+    ) -> str:
+        """
+        Generates a pre-signed URL authorizing a single ``method`` request on the
+        object with the given ``key``, valid for ``duration``.
+
+        The returned URL carries an Ed25519 signature in its query string, so the
+        recipient can perform the request without an auth token. It can be handed
+        to any HTTP client (``urllib``, ``requests``, a browser, ...); the URL is
```
Comment (lcian) [inline clients/python/src/objectstore_client/client.py]:
```
        to any HTTP client (``urllib``, ``requests``, a browser, ...); the URL is
```
I'm afraid this might not hold true in non-trivial cases that require encoding, as different HTTP clients might have subtle differences in encoding, and they probably don't have a way to force the usage of a given pre-parsed path/query string.

Comment (lcian) [reply clients/python/src/objectstore_client/client.py]: In https://github.com/getsentry/objectstore/pull/540#discussion_r3536931508 and our offline discussion with @jan-auer we decided that we want to sign on the **already encoded** components, so I'm committing to it.

Comment (lcian) [reply clients/python/src/objectstore_client/client.py]: Okay I did a test generating the pre-signed URLs in Python and hitting them with `reqwest` in Rust, turns out this is more robust than I thought, seems to work in practice:

```
  Pre-signed URL Manual Test Results

  Setup: objectstore server on localhost:18889, filesystem backend, auth enforced with the test Ed25519 keypair (test_kid).

  Phase 1: Python client PUT + presign generation

  - 28/28 keys PUT successfully (no failures)
  - 28/28 presigned GET URLs generated without error

  Phase 2: Python urllib.request fetch of presigned URLs

  - 28/28 OK — every presigned URL returned status 200 with correct body

  Phase 3: Rust reqwest fetch of the same presigned URLs

  - 28/28 OK — every presigned URL returned status 200 with correct body match

  Keys tested (all passed across all clients)

  From existing e2e suite (16):
  plain-key, with space, café-unicode, emoji-😀, plus+sign, sub!$'()*+,=delims, tilde~dot.key, 100%literal-percent, looks%20encoded, nested/path/segments, quote'apostrophe, at@sign, colon:separated, ampersand&inside, semi;colon, equals=sign

  Additional tricky keys (12):
  double  space, tabs\there, newline\nkey, mixed café 😀 +&=;, cjk-日本語, arabic-مرحبا, backslash\\key, curly{brace}key, square[bracket]key, pipe|key, caret^key, backtick`key

  Conclusion

  Zero failures. The pre-signed URL signing in the Python client produces URLs that verify correctly on the server when consumed by both Python's urllib and Rust's reqwest — including all encoding corner cases (unicode, multi-byte, special delimiters,
  percent-literal, whitespace variants, CJK, Arabic).
```

Comment (lcian) [reply clients/python/src/objectstore_client/client.py]: Tested also with `curl`:
```
⏺ 28/28 OK with curl — all presigned URLs work perfectly. That's now three independent HTTP clients (Python urllib, Rust reqwest, curl) all successfully consuming the same presigned URLs across all 28 encoding corner cases with zero failures.
```

File: clients/python/tests/test_e2e.py [hunk 2]
```diff
@@ -678,3 +680,151 @@ def put_part(part_number: int, data: bytes) -> CompletePart:
...
+ENCODING_CORNER_CASE_KEYS = [
+    "plain-key",
+    "with space",
+    "café-unicode",
+    "emoji-😀",
+    "plus+sign",
+    "sub!$'()*+,=delims",
+    "tilde~dot.key",
+    "100%literal-percent",  # `%li` isn't a valid escape, so both escape the `%`
+    # "looks%20encoded",  # diverges: quote -> %2520, urllib3 put -> %20
```
Comment (lcian) [inline clients/python/tests/test_e2e.py:809]: Fixed in https://github.com/getsentry/objectstore/pull/548

File: clients/python/src/objectstore_client/presign.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+#   - sub-delims: "!$&'()*+,;="
+#   - path extras: ":" "@" "/"  (the pchar set plus the "/" segment separator)
+#   - query also allows "?"
+# This is the same set urllib3 leaves unescaped, so the encoding is a fixed
+# point of the HTTP client that ultimately sends the URL and survives on the
+# wire verbatim. It is also byte-for-byte identical to the Rust client's
+# `percent_encoding` set (see `objectstore-types/src/presign.rs`), so the two
+# clients produce the same signed bytes for any key.
+_PATH_SAFE = "/:@!$&'()*+,;="
+_QUERY_SAFE = _PATH_SAFE + "?"
```
Comment (lcian) [inline clients/python/src/objectstore_client/presign.py:26]: I took a different approach initially which was using `urllib3`'s private functions (underscore-prefixed) to do the encoding.
This should be the same character set, and the encoding should be the same, the only difference is that we don't use a private API.

File: clients/python/src/objectstore_client/client.py [hunk 4]
```diff
@@ -446,6 +447,67 @@ def object_url(self, key: str) -> str:
...
+
+        if not isinstance(self._token, SecretKey):
+            raise ValueError("no secret key configured on this session")
+
+        if duration > presign.MAX_PRESIGN_DURATION:
+            raise ValueError(
+                f"duration {duration} exceeds the maximum of "
+                f"{presign.MAX_PRESIGN_DURATION}"
+            )
+        duration_secs = math.ceil(duration.total_seconds())
```
[the remark this answers is not in the document]
Comment (lcian) [reply clients/python/src/objectstore_client/client.py:486]: This is okay IMO.

File: clients/python/src/objectstore_client/client.py [hunk 5]
```diff
@@ -446,6 +447,67 @@ def object_url(self, key: str) -> str:
         """
         return self._make_url(key, full=True)
 
+    def presigned_object_url(
```
Comment (jan-auer) [inline clients/python/src/objectstore_client/client.py:450]: We also have the `object_url` function. Would it make sense to unify them in a way? For example:

```py
session.object_url(
    "foo/bar", 
    sign="GET",  # does not sign if not provided
    duration=timedelta(...), # optional, defaults to 1h
)
```

Just an alternative, current signature looks good to me.

File: clients/python/src/objectstore_client/client.py [hunk 6]
```diff
@@ -237,13 +239,12 @@ def mint_token(
...
-        generator.
+        Raises ``ValueError`` if no ``SecretKey`` is configured
+        or if any requested permission is not granted to the key.
         """
-        if not isinstance(self._token, TokenGenerator):
-            raise ValueError("no token generator configured on this session")
-        return self._token.sign_for_scope(
+        if not isinstance(self._token, SecretKey):
+            raise ValueError("no secret key configured on this session")
+        return self._token.token_for_scope(
```
Comment (jan-auer) [inline clients/python/src/objectstore_client/client.py:247]: Can you elaborate on why we need this changed?

File: clients/python/src/objectstore_client/client.py [hunk 7]
```diff
@@ -446,6 +447,67 @@ def object_url(self, key: str) -> str:
...
+        self,
+        method: Literal["GET", "HEAD"],
+        key: str,
+        duration: timedelta,
+    ) -> str:
+        """
+        Generates a pre-signed URL authorizing a single ``method`` request on the
+        object with the given ``key``, valid for ``duration``.
+
+        The returned URL carries an Ed25519 signature in its query string, so the
```
Comment (jan-auer) [inline clients/python/src/objectstore_client/client.py]: Let's be less concrete on what the signature is or has. it should be a opaque to the user so we can change out the implementation details at any point.

File: clients/python/src/objectstore_client/client.py [hunk 8]
```diff
@@ -446,6 +447,67 @@ def object_url(self, key: str) -> str:
...
+        recipient can perform the request without an auth token. It can be handed
+        to any HTTP client (``urllib``, ``requests``, a browser, ...); the URL is
+        already percent-encoded and must be transmitted verbatim.
+
+        Only ``GET`` and ``HEAD`` may be pre-signed. The permissions granted
+        are those configured server-side for the signing key. ``duration``
+        must not exceed one week.
+
+        Raises ``ValueError`` if no ``SecretKey`` is configured on this
```
Comment (jan-auer) [inline clients/python/src/objectstore_client/client.py]: _nit_: It might be useful to pull this information up, as this is a quite important restriction and precondition, not just a "by the way, this error can happen".

File: clients/python/README.md [hunk 9]
```diff
@@ -156,34 +156,57 @@ key = resumed.complete(new_parts + existing_parts)
...
-token = TokenGenerator(
+token = SecretKey(
     kid="my-service", secret_key="<private key>",
-).sign_for_scope("my_app", Scope(org=42, project=1337))
+).token_for_scope("my_app", Scope(org=42, project=1337))
 
 client = Client("http://localhost:8888", token=token)
 ```
 
+### Pre-signed URLs
```
Comment (jan-auer) [inline clients/python/README.md:187]: It would be useful to talk about redirects as a possible use case for this. And also call out that for external uploads, it's better to pass a scoped token and use the full client.

**Review by jan-auer — APPROVED** [review]

File: clients/python/src/objectstore_client/client.py (hunk 6 above)
Comment (lcian) [reply clients/python/src/objectstore_client/client.py:247]: It was called `sign_for_scope` but in reality it's encrypting a JWT.
`sign_for_scope` could be confused with the new pre-signed URL, so I wanted a clear distinction.

Comment (lcian) [reply clients/python/src/objectstore_client/client.py:247]: Similar reasoning for `TokenGenerator`.
It's actually a secret key that's not necessarily used to generate a token (JWT), but can also be used to sign for pre-signed URLs.

File: clients/python/src/objectstore_client/client.py (hunk 5 above)
Comment (lcian) [reply clients/python/src/objectstore_client/client.py:450]: I'd keep it like this for now.
But we should probably have a default on the duration, we currently don't, let me add that.

File: clients/python/README.md (hunk 9 above)
Comment (lcian) [reply clients/python/README.md:187]: @jan-auer Why is it that for external uploads it's better to pass a scoped token?

File: clients/python/src/objectstore_client/client.py (hunk 6 above)
Comment (lcian) [reply clients/python/src/objectstore_client/client.py:247]: I've kept the legacy aliases for now, so that this is not a breaking change.

File: clients/python/src/objectstore_client/presign.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+SUPPORTED_METHODS = ("GET", "HEAD")
+
+
+def encode_path(path: str) -> str:
+    """Percent-encodes a request path as it will appear on the wire.
+
+    Leaves the RFC 3986 unreserved set, sub-delims, and ``:`` ``@`` ``/``
+    unescaped (see :data:`_PATH_SAFE`); everything else is percent-encoded.
+    """
+    return quote(path, safe=_PATH_SAFE)
```
[the remark this answers is not in the document]
Comment (lcian) [reply clients/python/src/objectstore_client/presign.py:53]: Fixed in the follow-up PR.
